File: src/test/java/com/fwdekker/randomness/ui/JSpinnerRangeTest.java
package com.fwdekker.randomness.ui;

import com.fwdekker.randomness.ValidationException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import javax.swing.JSpinner;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;


/**
 * Unit tests for {@link JSpinnerRange}.
 */
final class JSpinnerRangeTest {
    private JSpinner min;
    private JSpinner max;


    @BeforeEach
    void beforeEach() {
        min = mock(JSpinner.class);
        max = mock(JSpinner.class);
    }


    @Test
    void testIllegalMaxRange() {
        assertThatThrownBy(() -> new JSpinnerRange(min, max, -37.20))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("maxRange must be a positive number.");
    }


    @Test
    void testRangeRelative() {
        when(min.getValue()).thenReturn(85.20);
        when(max.getValue()).thenReturn(-636.33);

        final JSpinnerRange range = new JSpinnerRange(min, max);

        assertThatThrownBy(() -> range.validate())
                .isInstanceOf(ValidationException.class)
                .hasMessage("The maximum should be no smaller than the minimum.");
    }

    @Test
    void testRangeSize() {
        when(min.getValue()).thenReturn(-1E53);
        when(max.getValue()).thenReturn(1E53);

        final JSpinnerRange range = new JSpinnerRange(min, max);

        assertThatThrownBy(() -> range.validate())
                .isInstanceOf(ValidationException.class)
                .hasMessage("The range should not exceed 1.0E53.");
    }

    @Test
    void testRangeSizeCustomRange() {
        when(min.getValue()).thenReturn(-794.90);
        when(max.getValue()).thenReturn(769.52);

        final JSpinnerRange range = new JSpinnerRange(min, max, 793.31);

        assertThatThrownBy(() -> range.validate())
                .isInstanceOf(ValidationException.class)
                .hasMessage("The range should not exceed 793.31.");
    }
}

File: src/test/kotlin/com/fwdekker/randomness/ui/JSpinnerRangeTest.kt
package com.fwdekker.randomness.ui

import io.kotest.core.spec.style.DescribeSpec
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.assertj.swing.edt.FailOnThreadViolationRepaintManager
import org.assertj.swing.edt.GuiActionRunner
import javax.swing.JSpinner


/**
 * Unit tests for the extension functions in `JSpinnerHelperKt`.
 */
object JSpinnerRangeTest : DescribeSpec({
    beforeContainer {
        FailOnThreadViolationRepaintManager.install()
    }


    describe("constructor") {
        it("throws an exception if the range is negative") {
            assertThatThrownBy { bindSpinners(createJSpinner(), createJSpinner(), -37.20) }
                .isInstanceOf(IllegalArgumentException::class.java)
                .hasMessage("maxRange must be a positive number.")
        }
    }

    describe("automatic range correction") {
        it("updates the minimum spinner if the maximum goes below its value") {
            val min = createJSpinner(150.38)
            val max = createJSpinner(244.54)
            bindSpinners(min, max)

            GuiActionRunner.execute { max.value = -284.85 }

            assertThat(min.value).isEqualTo(-284.85)
        }

        it("updates the maximum spinner if the minimum goes above its value") {
            val min = createJSpinner(-656.88)
            val max = createJSpinner(105.41)
            bindSpinners(min, max)

            GuiActionRunner.execute { min.value = 684.41 }

            assertThat(max.value).isEqualTo(684.41)
        }
    }
})


private fun createJSpinner() =
    GuiActionRunner.execute<JSpinner> { JSpinner() }

private fun createJSpinner(value: Double) =
    GuiActionRunner.execute<JSpinner> { JSpinner().also { it.value = value } }
